dev/PipsControl/PipsControl.cpp

```diff
@@ -0,0 +1,369 @@
+        SelectedPageIndex(numberOfPages - 1);
+    }
+    UpdateVerticalPips(numberOfPages, MaxDisplayedPages());
+    UpdateNavigationButtonVisualStates();
+}
+
+void PipsControl::OnSelectedPageIndexChange(const int oldValue)
+{
+    // If we don't have any pages, there is nothing we should do.
+    // Ensure that SelectedPageIndex will end up in the valid range of values
+    // Special case is NumberOfPages being 0, in that case, don't verify upperbound restrictions
+    if (SelectedPageIndex() > NumberOfPages() - 1 && NumberOfPages() > 0)
+    {
+        SelectedPageIndex(NumberOfPages() - 1);
+    }
+    else if (SelectedPageIndex() < 0)
+    {
+        SelectedPageIndex(0);
+    }
+    else {
+        // Now handle the value changes
+        m_lastSelectedPageIndex = oldValue;
+
+        // Fire value property change for UIA
+        if (const auto peer = winrt::FrameworkElementAutomationPeer::FromElement(*this).try_as<winrt::PipsControlAutomationPeer>())
+        {
+            winrt::get_self<PipsControlAutomationPeer>(peer)->RaiseSelectionChanged(m_lastSelectedPageIndex, SelectedPageIndex());
+        }
+        UpdateNavigationButtonVisualStates();
+        UpdateVerticalPips(NumberOfPages(), SelectedPageIndex());
+        RaiseSelectedIndexChanged();
+    }
+}
+
+void PipsControl::RaiseSelectedIndexChanged()
+{
+    const auto args = winrt::make_self<PipsControlSelectedIndexChangedEventArgs>(m_lastSelectedPageIndex, SelectedPageIndex());
+    m_selectedIndexChangedEventSource(*this, *args);
+}
+
+void PipsControl::MovePipIdentifierToElement(int index) {
+
+    if (NumberOfPages() != 0) {
+        if (const auto repeater = m_verticalPipsRepeater.get())
+        {
+            if (const auto element = repeater.TryGetElement(m_lastSelectedPageIndex).try_as<winrt::Button>()) {
+                element.Style(DefaultIndicatorStyle());
+            }
+            if (const auto element = repeater.GetOrCreateElement(index).try_as<winrt::Button>()) {
+                element.UpdateLayout();
+                element.Style(SelectedIndicatorStyle());
+                ScrollToCenterOfViewport(element);
+            }
+        }
+    }
+}
+
+void PipsControl::SetVerticalPipsSVMaxSize() {
+    auto pipHeight = unbox_value<double>(ResourceAccessor::ResourceLookup(*this, box_value(c_PipsControlButtonHeightPropertyName)));
+    auto numberOfPages = NumberOfPages() < 0 ? MaxDisplayedPages() : std::min(NumberOfPages(), MaxDisplayedPages());
+    auto scrollViewerHeight = pipHeight * numberOfPages;
+    if (const auto scrollViewer = m_verticalPipsScrollViewer.get()) {
+        m_verticalPipsScrollViewer.get().MaxHeight(scrollViewerHeight);
+    }
+}
+
+
+
+void PipsControl::UpdateVerticalPips(const int numberOfPages, const int maxDisplayedPages) {
+
+    auto pipsListSize = static_cast<int>(m_verticalPipsElements.Size());
+    auto const selectedIndex = SelectedPageIndex();
+
+    if (numberOfPages < pipsListSize) {
```

Comment:
>numberOfPages [](start = 8, length = 13)

I don't really get why we are doing this. Maybe a comment is needed

Reply:
When numberOfPages changed, if the number of pips we store is greater than than the new number of pages - we wipe them (better option is to only delete the "overflow), otherwise we keep adding :)